### ISSUE TITLE
Use of memory profiler

### ISSUE THREAD

**Issue by jubueche** [google/jax#4061]
Hello,
I have a memory leak that I identified to be in this section of my code:

```
process = psutil.Process(os.getpid())
print("1",process.memory_info().rss)
_, _, states_t = vmap(self._evolve_functional, in_axes=(None,None,0))(self._pack(), True,(inps,targets))
states_t["output_ts"].block_until_ready()
process = psutil.Process(os.getpid())
print("2",process.memory_info().rss)
```

and I would like to debug it using `jax.profiler.save_device_memory_profile("memory.prof")` but this function is not available. I am using Jax version 0.1.69 and jaxlib version 0.1.47 . Do I need to update them?
I am just running on CPU at the moment. How do you debug memory leaks on CPU (it only says for GPU and TPU in the doc).

Regarding the code:
You can see that I use `self._evolve_functional` as the callable in vmap. This is defined as follows:

```
@property
    def _evolve_functional(self):

        def evol_func(
            params: Params, is_learning: bool, sp_input_ts: np.ndarray,
        ) -> Tuple[np.ndarray, State, Dict[str, np.ndarray]]:
            # - Call the jitted evolution function for this layer
            # - Prepare the input
            if(is_learning):
                # - Unpack input and target
                (inps, I_target) = sp_input_ts
            else:
                inps = sp_input_ts
                I_target = None

            (
                new_state,
                Vmem_ts,
                spikes_ts,
                r_ts,
                output_ts,
                PInv,
                w_out
            ) = _evolve_jit_FORCE(
                state0 = self.state,
                w_in = self.w_in,
                w_rec = self.w_rec,
                w_out = params["w_out"],
                PInv = params["PInv"],
                tau_mem = self.tau_mem,
                tau_syn = self.tau_syn,
                bias = self.bias,
                t_ref = self.t_ref,
                v_thresh = self.v_thresh,
                v_reset = self.v_reset,
                noise_std = self.noise_std,
                I_input = inps,
                key = self._rng_key,
                dt = self.dt,
                alpha = self.alpha,
                is_learning = is_learning,
                E = self.E,
                I_target = I_target
            )

            # - Return the outputs from this layer, and the final layer state
            states_t = {
                "Vmem": Vmem_ts,
                "r": r_ts,
                "output_ts": output_ts,
                "PInv": PInv,
                "w_out": w_out
            }
            return spikes_ts, new_state, states_t

        # - Return the evolution function
        return evol_func
```

So it basically just calls the function `_evolve_jit_FORCE` with the given arguments.
The function `_evolve_jit_FORCE` is defined as follows:

```
def _evolve_jit_FORCE(state0,
                        w_in,
                        w_rec,
                        w_out,
                        PInv,
                        tau_mem,
                        tau_syn,
                        bias,
                        t_ref,
                        v_thresh,
                        v_reset,
                        noise_std,
                        I_input,
                        key,
                        dt,
                        alpha,
                        is_learning,
                        E,
                        I_target):

        tau_syn = tau_syn.reshape((-1,1))
        tau_mem = tau_mem.reshape((-1,1))
        v_thresh = v_thresh.reshape((-1,1))
        v_reset = v_reset.reshape((-1,1))
        bias = bias.reshape((-1,1))
        tau_mem_kernel = (dt / tau_mem)
        exp_tau_syn_kernel = np.exp(-dt / tau_syn)

        state = state0
        training_step = 50

        @jit
        def forward_train(train_state, input_target_pair):
            (w_out, PInv, state) = train_state
            (I_input_ts, I_target_ts) = input_target_pair
            (w_out,state), (_, _, _, z) = forward((w_out, state), I_input_ts)
            # if(state["t"] % training_step == 0): # - TODO
            err = z - I_target_ts.reshape((-1,1))
            cd = PInv @ state["r"]
            w_out = w_out - (cd @ err.T)
            PInv = PInv - (cd @ cd.T)/ (1 + state["r"].T @ cd)
            return (w_out, PInv, state), (state["Vmem"],state["spikes"],state["r"], z)

        @jit
        def forward(evolve_state, I_input_ts):
            (w_out, state) = evolve_state
            z = w_out.T @ state["r"]
            I = state["Isyn"] + E @ z + (I_input_ts @ w_in).reshape((-1,1)) + bias
            dv = ((dt*state["t"])>(state["tlast"] + t_ref)).astype(np.int32) * (-state["Vmem"]+I)
            state["Vmem"] = state["Vmem"] + tau_mem_kernel * dv
            state["spikes"] = (state["Vmem"]>=v_thresh).astype(np.float32)
            I_rec = (w_rec @ state["spikes"]).reshape((-1,1))
            state["tlast"] = state["tlast"] + (dt*state["t"] -state['tlast']) * state["spikes"]
            state["Isyn"] = state["Isyn"]*exp_tau_syn_kernel + I_rec/tau_syn
            state["r"] = state["r"] * exp_tau_syn_kernel + state["spikes"]/tau_syn
            state["Vmem"] = state["Vmem"] + (30 - state["Vmem"]) * state["spikes"]
            state["Vmem"] = state["Vmem"] + (v_reset - state["Vmem"]) * state["spikes"]
            state["t"] += 1
            return (w_out, state), (state["Vmem"],state["spikes"],state["r"], z)

        # - Create membrane potential noise trace
        num_timesteps = I_input.shape[0]
        _, subkey = rand.split(key)
        noise_ts = noise_std * rand.normal(
            subkey, shape=(num_timesteps, I_input.shape[1])
        )

        I_input = I_input + noise_ts

        if(not is_learning):
            (w_out,state), (Vmem_ts, spikes_ts, r_ts, output_ts) = scan(forward,
                            (w_out,state0),
                            I_input)
        else:
            (w_out,PInv,state), (Vmem_ts, spikes_ts, r_ts, output_ts) = scan(forward_train,
                            (w_out,PInv,state0),
                            (I_input, I_target))

        return state, Vmem_ts, spikes_ts, r_ts, output_ts, PInv, w_out
```

which as you can see is not even jitted at the moment, because I figured the part that counts is the forward and forward train methods.

So to summarize my problems:
There is a memory leak there somewhere and I don't know how to debug it properly.

Thanks for your help. Jax is a really nice tool!

All the best,
Julian

**Comment by jekbradbury** [comment]
You do need to update your jax and jaxlib to use the memory profiler. I believe it supports CPU in addition to GPU and TPU.

**Comment by jubueche** [comment]
Thanks a lot, I can profile now :)
I am having a little bit of trouble identifying the problem though. I have attached the profile for the program up until
```
process = psutil.Process(os.getpid())
print("1",process.memory_info().rss)
_, _, states_t = vmap(self._evolve_functional, in_axes=(None,None,0))(self._pack(), True,(inps,targets))
states_t["output_ts"].block_until_ready()
process = psutil.Process(os.getpid())
print("2",process.memory_info().rss)
```
which looks like this
[profile002.pdf](https://github.com/google/jax/files/5083509/profile002.pdf)

I have also attached a profile using --diff_base between two consecutive iterations:
[profile001.pdf](https://github.com/google/jax/files/5083615/profile001.pdf)

Judging from the second figure I guess the memory leak comes from the function ```get_data()```. Is that right?

If so, these are the call snippets: Maybe you see something odd.

```get_data()``` makes a call to ```vmap``` in this snippet:

```
def get_data(self, filtered_batch):
        batched_rate_output, _, states_t = vmap(self.rate_layer._evolve_functional, in_axes=(None, None, 0))(self.lr_params, self.lr_state, filtered_batch)
        batched_res_inputs = states_t["res_inputs"]
        batched_res_acts = states_t["res_acts"]
        return batched_rate_output, batched_res_inputs, batched_res_acts
```
It uses the class property ```_evolve_functional``` for vmap. This functional is:

```
@property
    def _evolve_functional(self):

        def evol_func(
            params: Params, state: State, inputs: np.ndarray,
        ):
            # - Call the jitted evolution function for this layer
            (
                new_state,
                res_inputs,
                rec_inputs,
                res_acts,
                outputs,
                key1,
            ) = self._evolve_jit(
                state,
                params["w_in"],
                params["w_recurrent"],
                params["w_out"],
                params["bias"],
                params["tau"],
                inputs,
                self._noise_std,
                self._rng_key,
                self._dt,
            )

            # - Maintain RNG key, if not under compilation
            if not isinstance(key1, jax.core.Tracer):
                self._rng_key = key1

            # - Return the outputs from this layer, and the final layer state
            states_t = {
                "res_inputs": res_inputs,
                "rec_inputs": rec_inputs,
                "res_acts": res_acts,
            }
            return outputs, new_state, states_t

        # - Return the evolution function
        return evol_func
```
which is a wrapper for the function ```_evolve_jit``` which is defined as
```
self._evolve_jit = _get_rec_evolve_jit(self._H)
```
where ```self._H``` is just some activation function (also jitted) and ```_get_rec_evolve_jit``` is defined as

```
def _get_rec_evolve_jit(
    H: Callable[[float], float]
) -> Callable[
    [
        np.ndarray,
        np.ndarray,
        np.ndarray,
        np.ndarray,
        np.ndarray,
        np.ndarray,
        np.ndarray,
        float,
        List[int],
        float,
    ],
    Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray, Any],
]:
    """
    Return a compiled raw reservoir evolution function

    :param Callable[[float], float] H:   Neuron activation function
    :return Callable:     f(x0, w_in, w_recurrent, w_out, bias, tau, inputs, noise_std, key, dt) -> (x, res_inputs, rec_inputs, res_acts, outputs)
    """

    #@jit
    def rec_evolve_jit(
        x0: np.ndarray,
        w_in: FloatVector,
        w_recurrent: FloatVector,
        w_out: FloatVector,
        bias: FloatVector,
        tau: FloatVector,
        inputs: np.ndarray,
        noise_std: float,
        key,
        dt: float,
    ) -> Tuple[np.ndarray, np.ndarray, np.ndarray, np.ndarray, np.ndarray, Any]:
        """
        Compiled recurrent evolution function

        :param np.ndarray x0:       Initial state of forced layer [N]
        :param np.ndarray w_in:     Input weights [IxN]
        :param np.ndarray w_recurrent:     Recurrent weights [NxN]
        :param np.ndarray w_out:    Output weights [NxO]
        :param np.ndarray bias:     Bias values of reservoir units [N]
        :param np.ndarray tau:      Time constants of reservoir units [N]
        :param np.ndarray inputs:   Input time series [TxN]
        :param np.ndarray force:    Driving time series injected into reservoir units instead of recurrent activity [TxN]
        :param float noise_std:     Standard deviation of noise injected into reservoir units
        :param Any key:             Jax RNG key to use in noise generation
        :param float dt:            Time step for forward Euler solver

        :return:    (x, res_inputs, rec_inputs, res_acts, outputs, key)
                x:          np.ndarray State of
                res_inputs: np.ndarray Time series of weighted external inputs to each reservoir unit [TxN]
                rec_inputs: np.ndarray Time series of recurrent inputs to each reservoir unit [TxN]
                res_acts:   np.ndarray Time series of reservoir unit activities [TxN]
                outputs:    np.ndarray Time series of output layer values [TxO]
                key:        Any        New RNG key
        """
        # - Pre-compute dt/tau
        dt_tau = dt / tau

        # - Reservoir state step function (forward Euler solver)
        def reservoir_step(x, inps):
            """
            reservoir_step() - Single step of recurrent reservoir

            :param x:       np.ndarray Current state of reservoir units
            :param inps:    np.ndarray Inputs to each reservoir unit for the current step

            :return:    xnext, (rec_input, activation)
            """
            inp, rand = inps
            activation = H(x)
            rec_input = np.dot(activation, w_recurrent)
            dx = dt_tau * (-x + inp + bias + rand + rec_input)

            return x + dx, (rec_input, activation)

        # - Evaluate passthrough input layer
        res_inputs = np.dot(inputs, w_in)

        # - Compute random numbers for reservoir noise
        key1, subkey = rand.split(key)
        noise = noise_std * rand.normal(subkey, shape=(inputs.shape[0], np.size(x0)))

        # - Use `scan` to evaluate reservoir
        x, (rec_inputs, res_acts) = scan(reservoir_step, x0, (res_inputs, noise))

        # - Evaluate passthrough output layer
        outputs = np.dot(res_acts, w_out)

        return x, res_inputs, rec_inputs, res_acts, outputs, key1

    return rec_evolve_jit
```

**Comment by jubueche** [comment]
I disbaled caching for ```scan``` and now the memory is leaking at a much slower pace. Is it possible that this ```scan```:

```
        tau_syn = tau_syn.reshape((-1,1))
        tau_mem = tau_mem.reshape((-1,1))
        v_thresh = v_thresh.reshape((-1,1))
        v_reset = v_reset.reshape((-1,1))
        bias = bias.reshape((-1,1))
        tau_mem_kernel = (dt / tau_mem)
        exp_tau_syn_kernel = np.exp(-dt / tau_syn)

        state = state0
        training_step = 50

        def forward_train(train_state, input_target_pair):
            (w_out, PInv, state) = train_state
            (I_input_ts, I_target_ts) = input_target_pair
            (w_out,state), (_, _, _, z) = forward((w_out, state), I_input_ts)
            # if(state["t"] % training_step == 0): # - TODO
            err = z - I_target_ts.reshape((-1,1))
            cd = PInv @ state["r"]
            w_out = w_out - (cd @ err.T)
            PInv = PInv - (cd @ cd.T)/ (1 + state["r"].T @ cd)
            return (w_out, PInv, state), (state["Vmem"],state["spikes"],state["r"], z)

        def forward(evolve_state, I_input_ts):
            (w_out, state) = evolve_state
            z = w_out.T @ state["r"]
            I = state["Isyn"] + E @ z + (I_input_ts @ w_in).reshape((-1,1)) + bias
            dv = ((dt*state["t"])>(state["tlast"] + t_ref)).astype(np.int32) * (-state["Vmem"]+I)
            state["Vmem"] = state["Vmem"] + tau_mem_kernel * dv
            state["spikes"] = (state["Vmem"]>=v_thresh).astype(np.float32)
            I_rec = (w_rec @ state["spikes"]).reshape((-1,1))
            state["tlast"] = state["tlast"] + (dt*state["t"] -state['tlast']) * state["spikes"]
            state["Isyn"] = state["Isyn"]*exp_tau_syn_kernel + I_rec/tau_syn
            state["r"] = state["r"] * exp_tau_syn_kernel + state["spikes"]/tau_syn
            state["Vmem"] = state["Vmem"] + (30 - state["Vmem"]) * state["spikes"]
            state["Vmem"] = state["Vmem"] + (v_reset - state["Vmem"]) * state["spikes"]
            state["t"] += 1
            return (w_out, state), (state["Vmem"],state["spikes"],state["r"], z)

        # - Create membrane potential noise trace
        num_timesteps = I_input.shape[0]
        _, subkey = rand.split(key)
        noise_ts = noise_std * rand.normal(
            subkey, shape=(num_timesteps, I_input.shape[1])
        )

        I_input = I_input + noise_ts

        if(not is_learning):
            (w_out,state), (Vmem_ts, spikes_ts, r_ts, output_ts) = scan(forward,
                            (w_out,state0),
                            I_input)
        else:
            (w_out,PInv,state), (Vmem_ts, spikes_ts, r_ts, output_ts) = scan(forward_train,
                            (w_out,PInv,state0),
                            (I_input, I_target))
```
is being re-compiled the whole time because of the values that I am not passing, e.g. ```bias```?
How can I check if my functions are being re-compiled?

Thanks a lot :)

**Comment by jubueche** [comment]
Ok, the leak disappears when I jit ```_evolve_jit_FORCE``` using ```@jax.partial(jax.jit, static_argnums=(16,))```. Any idea why that could be the case?
